### PR TITLE
VBLOCKS-1800 iOS incoming call flow refactor

### PR DESCRIPTION
**Pull request by bobiechen-twilio**
## Submission Checklist

 - [ ] Updated the `CHANGELOG.md` to reflect any **feature**, **bug fixes**, or **known issues** made in the source code
 - [ ] Tested code changes and observed expected behavior in the example app
 - [ ] Performed a visual inspection of the `Files changed` tab prior to submitting the pull request for review to ensure proper usage of the style guide

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

## Description

- [ ] Link issue links

### Problem description

- The SDK (iOS layer) intercepts all the PushKit notifications and the applications have no control over it. Some application already have PushKit logic implemented but now all notifications are redirected to the RN SDK.
- The VoIP push notification is reported to CallKit by the SDK to avoid the policy violation termination. The applications have no way of controlling the caller handle/name being reported to CallKit.

Screenshot 2023-06-26 at 3 40 52 PM

### Refactor solution

- Allow applications to opt out and use their own iOS PushKit implementation. The applications will be able to receive and control the push notifications they receive, especially if they have other modules also depending on push notifications. The `Voice.initializePushRegistry()` method will allow the app to delegate this back to the SDK.
- A new method `Voice.handlePushNotification()` for the applications to pass the push payload to the SDK, if the platform push notification module is implemented outside of the SDK. `CallInvite` event will be returned as it is already doing.
- A new method `Voice.reportNewIncomingCall()` for the iOS applications to report to CallKit, with the flexibility to determine the caller name instead of the SDK using the `from` value of the call invite. For Android applications, this is not required.

### Issues reported by developers

- #175 

## Breakdown

- [Bulleted summary of changes]
- [eg. Updated documentation]
- [eg. Added audio device feature]

## Validation

-

### CODE REVIEW

**Comment by bobiechen-twilio** [comment]
from the discussion:
- document the values in the call invite
- implement voice_handlePushNotification() for Android
  - [ ] create a ticket for Android specific task
- leave CallInvite.reportNewIncomingCall() for later

File: api/voice-react-native-sdk.api.md [hunk 1]
```diff
@@ -900,6 +900,7 @@ export class Voice extends EventEmitter {
     getCalls(): Promise<ReadonlyMap<Uuid, Call>>;
     getDeviceToken(): Promise<string>;
     getVersion(): Promise<string>;
+    initializePushRegistry(): Promise<void>;
```
Comment (charliesantos) [inline api/voice-react-native-sdk.api.md:903]: Do we need to update IDL?

File: ios/docs/applications-own-pushkit-handler.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+To use your existing iOS PushKit delegate module, a few steps need to be done by the application:
+
+- Provide the PushKit device token (in `NSData`) to the React Native Voice SDK
+- Provide the PushKit notification payload (in `NSDictionary`) to the React Native Voice SDK
+
+### PushKit device token updated
+
+Upon getting the PushKit device token update, post a notification and include the device token (in the form of `NSData`) in the `userInfo`. The device token will later be used when calling the `Voice.register()` method.
```
Comment (mhuynh5757) [inline ios/docs/applications-own-pushkit-handler.md:19]:
```suggestion
Upon receiving the PushKit device token update, post a notification and include the device token (in the form of `NSData`) in the `userInfo`. The device token will later be used when calling the `Voice.register()` method.
```

File: ios/docs/applications-own-pushkit-handler.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                                                                     @"incomingPushPayload": payload.dictionaryPayload}];
+    }
+
+    completion();
+}];
+```
+
+A `CallInvite` event will be raised to the application so the user can accept or reject the call invite.
+
+### Reporting new incoming call to iOS CallKit
```
Comment (mhuynh5757) [inline ios/docs/applications-own-pushkit-handler.md:56]:
```suggestion
### Reporting a new incoming call to iOS CallKit
```

**Review by mhuynh5757 — APPROVED** [review]

**Review by mhuynh5757 — CHANGES_REQUESTED** [review]
Apologies, misclicked the Approve button. I had minor comments, but I think they should be addressed.

File: ios/TwilioVoicePushRegistry.m [hunk 4]
```diff
@@ -14,10 +14,11 @@
 #import "TwilioVoiceReactNativeConstants.h"
 
 NSString * const kTwilioVoicePushRegistryNotification = @"TwilioVoicePushRegistryNotification";
+NSString * const kTwilioVoicePushRegistryEventType = @"type";
 NSString * const kTwilioVoicePushRegistryNotificationDeviceTokenUpdated = @"deviceTokenUpdated";
-NSString * const kTwilioVoicePushRegistryNotificationDeviceTokenKey = @"deviceToken";
-NSString * const kTwilioVoicePushRegistryNotificationCallInviteKey = @"userInfoCallInvite";
-NSString * const kTwilioVoicePushRegistryNotificationCancelledCallInviteKey = @"userInfoCancelledCallInvite";
+NSString * const kTwilioVoicePushRegistryNotificationDeviceToken = @"deviceToken";
```
[the remark this answers is not in the document]
Comment (bobiechen-twilio) [reply ios/TwilioVoicePushRegistry.m:19]: The string literal `deviceToken` is only used in the iOS so I guess it's fine to keep it in the iOS layer.

File: ios/TwilioVoiceReactNative.m [hunk 5]
```diff
@@ -437,6 +416,13 @@ - (void)sendEventWithName:(NSString *)eventName body:(id)body {
     }
 }
 
+RCT_EXPORT_METHOD(voice_initializePushRegistry:(RCTPromiseResolveBlock)resolve
+                  rejecter:(RCTPromiseRejectBlock)reject)
+{
+    [self initializePushRegistry];
```
[the remark this answers is not in the document]
Comment (bobiechen-twilio) [reply ios/TwilioVoiceReactNative.m:422]: Good point, but the iOS PushKit initialization seems to be a fire and forget op. Doesn't seem to be a way to know things went wrong.

[the remark this answers is not in the document]
Comment (mhuynh5757) [reply ios/TwilioVoiceReactNative.m:422]: LOL, okay well that's concerning but nothing we can do about it 🤷

File: src/Voice.tsx [hunk 6]
```diff
@@ -814,14 +814,16 @@ export class Voice extends EventEmitter {
 
   /**
    * Initialize a Push Registry instance inside the SDK for handling
-   * PushKit device token update and receiving push notifications.
+   * PushKit device token updates and receiving push notifications.
    *
    * @remarks
+   Unsupported platforms:
```
Comment (mhuynh5757) [inline src/Voice.tsx:820]: Sorry, I missed the prefix `*` characters in the last review.

**Review by mhuynh5757 — CHANGES_REQUESTED** [review]

**Review by mhuynh5757 — APPROVED** [review]

**Pull request merged**